src/libraries/System.Private.CoreLib/src/System/Reflection/InvokerEmitUtil.cs

```diff
@@ -0,0 +1,331 @@
+                    il.Emit(OpCodes.Stloc_S, tmp);
+                    il.Emit(OpCodes.Ldloc_S, tmp);
+
+                    il.Emit(OpCodes.Ldobj, typeof(object));
+                    il.Emit(OpCodes.Unbox, parameterType);
+                    il.Emit(OpCodes.Ldobj, parameterType);
+                }
+                else if (parameterType.IsPointer)
+                {
+                    LocalBuilder tmp = il.DeclareLocal(typeof(IntPtr).MakeByRefType());
+                    il.Emit(OpCodes.Stloc_S, tmp);
+                    il.Emit(OpCodes.Ldloc_S, tmp);
+                    il.Emit(OpCodes.Ldobj, typeof(IntPtr));
+                }
+                else
+                {
+                    LocalBuilder tmp = il.DeclareLocal(parameterType.MakeByRefType());
+                    il.Emit(OpCodes.Stloc_S, tmp);
+                    il.Emit(OpCodes.Ldloc_S, tmp);
+                    il.Emit(OpCodes.Ldobj, parameterType);
+                }
+
+                // Move to the next argument.
+                if (i < parameters.Length - 1)
+                {
+                    il.Emit(OpCodes.Ldloc_S, local_pArg);
+                    il.Emit(OpCodes.Sizeof, typeof(IntPtr));
+                    il.Emit(OpCodes.Add);
+                    il.Emit(OpCodes.Stloc_S, local_pArg);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Update any nullables that were passed by reference.
+        /// </summary>
+        private static void UpdateNullables(
+            ILGenerator il,
+            ParameterInfo[] parameters,
+            NullableRefInfo[] byRefNullables,
+            int byRefNullablesCount)
+        {
+            for (int i = 0; i < byRefNullablesCount; i++)
+            {
+                NullableRefInfo info = byRefNullables[i];
+
+                RuntimeType parameterType = (RuntimeType)parameters[info.ParameterIndex].ParameterType;
+                Debug.Assert(parameterType.IsByRef);
+
+                RuntimeType? elementType = (RuntimeType)parameterType.GetElementType()!;
+                Debug.Assert(elementType.IsNullableOfT);
+
+                // Get the original byref location.
+                il.Emit(OpCodes.Ldc_I4_S, info.ParameterIndex);
+                il.Emit(OpCodes.Conv_I);
+                il.Emit(OpCodes.Sizeof, typeof(IntPtr));
+                il.Emit(OpCodes.Mul);
+                il.Emit(OpCodes.Ldarg_2);
+                il.Emit(OpCodes.Add);
+                il.Emit(OpCodes.Ldind_I);
+
+                // Get the Nullable<T>& value and update the original.
+                il.Emit(OpCodes.Ldloc_S, info.LocalIndex);
+                il.Emit(OpCodes.Ldobj, elementType);
+                il.Emit(OpCodes.Box, elementType);
+                il.Emit(OpCodes.Stind_Ref);
+            }
+        }
+
+        private static void Invoke(ILGenerator il, MethodBase method)
+        {
+            // todo: once we return the value without boxing add support for OpCodes.Tailcall for perf.
```

Comment:
The JIT will do a tailcall for you when it sees it as beneficial.

Explicit tailcall. prefix can both help or hurt performance depending on the platform calling convention details . It is better to leave it to the JIT to figure out whether tailcall is a good perf optimization.